enhancements/network/requesting-staticips-for-vms-being-migrated-into-primary-l2-udns-using-MTV.md

```diff
@@ -0,0 +1,849 @@
+OVN-Kubernetes provides BGP for UDNs, we should follow-up with ensuring this
+feature works for UDNs as well.
+
+## Proposal
+
+This proposal to import VMs from other virtualization platforms while
+preserving their networking configuration will impact three different OpenShift
+components:
+- MTV (Migration Toolkit for Virtualization)
+- KubeVirt's ipam-extensions
+- OVN-Kubernetes
+
+We will elaborate the overall flow first, before digging into details in each
+of the components. There will be two sub-sections: one about preserving the
+[IP addresses](#preserving-the-original-vm-ip-address),
+another about preserving the
+[gateway configuration](#preserving-the-vm-gateway).
+MTV already introspects the VM on the source cluster, and templates the VM with
+its original MAC addresses.
+
+### Current limitations
+
+- MTV does **not** support importing VMs with IPv6 addresses. Whatever support
+  for IPv6 we come up with, will probably be limited by this.
+
+### Preserving the original VM IP address
+
+MTV currently introspects the VM to learn the MAC addresses of the interfaces,
+and templates the VM which will be created in OpenShift Virtualization with the
+required MAC address. We would need MTV to somehow also figure out what IP
+addresses are on the aforementioned interfaces.
+
+To allow OpenShift Virtualization to import VMs - preserving their MAC and IP
+addresses - we need to have both addresses stored somehow in the system (*at
+least* for the VMs which will be migrated into OpenShift Virt).
+
+The simplest alternative (which doesn't require any API changes) is to have MTV
+template the OpenShift Virt VM with the required information - the MAC and IP
+addresses for the primary UDN attachment.
+
+The MAC address is currently defined in the VM spec, it is an attribute of the
+interfaces in the `spec.template.spec.domain.devices.interfaces` array.
+Unfortunately, there isn't a place in the KubeVirt API to request the IP
+addresses for the interface (or network attachment). Given we're already past
+the 4.20 KubeVirt feature freeze, we propose templating the VM with an
+annotation where the interface's IPs will be defined. This annotation
+will be KubeVirt specific, and will be consumed by the KubeVirt
+`ipam-extensions` project, which will mutate the launcher pod to customize the
+cluster default network attachment, by using the multus default network
+feature. The multus default network annotation's value is a
+network-selection-element, which we can use to instruct the pod's IPs and MAC
+addresses. The annotation on the VMs will look like:
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: vm-server
+  namespace: blue
+spec:
+  runStrategy: Always
+  template:
+    metadata:
+      annotations:
+        # the annot below defines the IPs for this logical network of the imported VM
+        network.kubevirt.io/ip-addresses: '{"iface1": ["192.168.0.1/24", "fd23:3214::123/64"]}'
```

Comment:
Since this annotation is only passed-through KubeVirt (and not owned by it), please consider using another URL.
Same bellow.

Reply:
It's passed through kubevirt/kubevirt to kubevirt/ipam-extensions https://github.com/kubevirt/ipam-extensions/pull/105.

network.kubevirt.io is used by CNAO and its components https://github.com/kubevirt/cluster-network-addons-operator